lib/control-connection.js

```diff
@@ -773,6 +812,74 @@ ControlConnection.prototype.reset = function (callback) {
+ * not support clusters with nodes that have versions that are more than one major
+ * version away from each other.
+ *
+ * @param {Number} version Maximum version to consider.
+ * @param {Array.<Host>} hosts The hosts to determine highest protocol version from.
+ * @return {Number} Highest supported protocol version among hosts.
+ */
+ControlConnection.prototype._highestCommonProtocolVersion = function(version, hosts) {
+  let maxVersion = version;
+  // whether or not protocol v3 is required (nodes detected that don't support < 3).
+  let v3Requirement = false;
+  // track the common protocol version >= v3 in case we encounter older versions.
+  let maxVersionWith3OrMore = version;
+  const self = this;
+  hosts.forEach(h => {
+    if (!h.cassandraVersion || h.cassandraVersion.length === 0) {
+      self.log('warning', 'Encountered host ' + h.address + ' with no cassandra version,' +
+        ' skipping as part of protocol version evaluation');
+      return;
+    }
+
+    try {
+      const cassandraVersion = VersionNumber.parse(h.cassandraVersion);
+      self.log('verbose', 'Encountered host ' + h.address + ' with cassandra version ' + cassandraVersion);
+      if (cassandraVersion.compare(V3_0_0) >= 0) {
+        // Anything 3.0.0+ has a max protocol version of V4 and requires at least V3.
+        v3Requirement = true;
+        maxVersion = Math.min(types.protocolVersion.v4, maxVersion);
+        maxVersionWith3OrMore = maxVersion;
+      } else if (cassandraVersion.compare(V2_2_0) >= 0) {
+        // Cassandra 2.2.x has a max protocol version of V4.
+        maxVersion = Math.min(types.protocolVersion.v4, maxVersion);
+        maxVersionWith3OrMore = maxVersion;
+      } else if (cassandraVersion.compare(V2_1_0) >= 0) {
+        // Cassandra 2.1.x has a max protocol version of V3.
+        maxVersion = Math.min(types.protocolVersion.v3, maxVersion);
+        maxVersionWith3OrMore = maxVersion;
+      } else if (cassandraVersion.compare(V2_0_0) >= 0) {
+        // Cassandra 2.0.x has a max protocol version of V2.
+        maxVersion = Math.min(types.protocolVersion.v2, maxVersion);
+      } else {
+        // Anything else is < 2.x and requires protocol version V1.
+        maxVersion = types.protocolVersion.v1;
+      }
+    } catch (e) {
+      self.log('warning', 'Encountered host ' + h.address + ' with unparseable cassandra version ' + h.cassandraVersion
+        + ' skipping as part of protocol version evaluation');
+    }
+  });
+
+  if (v3Requirement && maxVersion < types.protocolVersion.v3) {
+    self.log('warning', 'Detected hosts with maximum protocol version of ' + maxVersion +
+      ' but there are some hosts that require at least version 3, will not be able to connect to these older hosts');
+    maxVersion = maxVersionWith3OrMore;
+  }
+
+  self.log('verbose', 'Resolved protocol version ' + maxVersion + ' as the highest common protocol version among hosts');
```

Comment:
As the decimal representation of the protocol version could be misleading, we can use the hex representation instead: `'Resolved protocol version 0x' + maxVersion.toString(16)`